optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/rules/QuadGroupBy2Map1CollectFastMutator.java

```diff
@@ -0,0 +1,77 @@
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.common.rules;
+
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.singletonList;
+import static org.drools.model.DSL.accFunction;
+import static org.drools.model.PatternDSL.from;
+import static org.drools.model.PatternDSL.groupBy;
+import static org.drools.model.PatternDSL.pattern;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.drools.model.PatternDSL;
+import org.drools.model.Variable;
+import org.drools.model.view.ViewItem;
+import org.optaplanner.core.api.function.QuadFunction;
+import org.optaplanner.core.api.score.stream.quad.QuadConstraintCollector;
+import org.optaplanner.core.impl.score.stream.drools.common.BiTuple;
+import org.optaplanner.core.impl.score.stream.drools.quad.DroolsQuadAccumulateFunction;
+
+final class QuadGroupBy2Map1CollectFastMutator<A, B, C, D, NewA, NewB, NewC> extends AbstractQuadGroupByMutator {
+
+    private final QuadFunction<A, B, C, D, NewA> groupKeyMappingA;
+    private final QuadFunction<A, B, C, D, NewB> groupKeyMappingB;
+    private final QuadConstraintCollector<A, B, C, D, ?, NewC> collectorC;
+
+    public QuadGroupBy2Map1CollectFastMutator(QuadFunction<A, B, C, D, NewA> groupKeyMappingA,
+            QuadFunction<A, B, C, D, NewB> groupKeyMappingB, QuadConstraintCollector<A, B, C, D, ?, NewC> collectorC) {
+        this.groupKeyMappingA = groupKeyMappingA;
+        this.groupKeyMappingB = groupKeyMappingB;
+        this.collectorC = collectorC;
+    }
+
+    @Override
+    public AbstractRuleAssembler apply(AbstractRuleAssembler ruleAssembler) {
+        ruleAssembler.applyFilterToLastPrimaryPattern();
+        Variable<A> inputA = ruleAssembler.getVariable(0);
+        Variable<B> inputB = ruleAssembler.getVariable(1);
+        Variable<C> inputC = ruleAssembler.getVariable(2);
+        Variable<D> inputD = ruleAssembler.getVariable(3);
+        Variable<BiTuple<NewA, NewB>> groupKey = ruleAssembler.createVariable(BiTuple.class, "groupKey");
+        Variable<NewC> output = ruleAssembler.createVariable("output");
+        ViewItem groupByPattern = groupBy(getInnerAccumulatePattern(ruleAssembler), inputA, inputB, inputC, inputD,
+                groupKey, (a, b, c, d) -> new BiTuple<>(groupKeyMappingA.apply(a, b, c, d),
+                        groupKeyMappingB.apply(a, b, c, d)),
+                accFunction(() -> new DroolsQuadAccumulateFunction<>(collectorC)).as(output));
+        List<ViewItem> newFinishedExpressions = new ArrayList<>(ruleAssembler.getFinishedExpressions());
+        newFinishedExpressions.add(groupByPattern); // The last pattern is added here.
+        Variable<NewA> newA = ruleAssembler.createVariable("newA", from(groupKey, k -> k.a));
+        Variable<NewB> newB = ruleAssembler.createVariable("newB", from(groupKey, k -> k.b));
+        Variable<NewC> newC = ruleAssembler.createVariable("newB", from(output));
```

Comment:
Is the variable name "newB" by any chance a copy&paste mistake?

Reply:
Good catch. That said, this is not a bug, since `createVariable(...)` makes the var names unique anyway.